### ISSUE TITLE
support hotplug/hot-unplug VFIO Device in container running state

### ISSUE THREAD

**Issue by soulseen** [kata-containers/kata-containers#5172]
**Is your feature request related to a problem? Please describe.**

In some cases, we need to hot plug/unplug GPUs Device(or other VFIO devices) dynamically into kata container when it is running, to implement the GPUs dynamically assign between different kata containers.

**Describe the solution you'd like**

AFAIK we are already implemented the [DeviceManager](https://github.com/kata-containers/kata-containers/blob/6b0dc60dda37251f7c4b25ff1cfb10ed11dc88e4/src/runtime/virtcontainers/device/api/interface.go#L89) interface. so,  I think we could add a new Kata-specific sub-command like:
```
kata-runtime hotplug $container $device
```
that would allow users to hotplug or hot-unplug VFIO Devices.

**Comment by fengwang666** [comment]
Exposing an interface through kata-runtime is potentially an option. Can you explain a bit more about the use case? Sounds like the GPU hotplug/unplug lifecycle is different from the pod/containers? How would your container application handle the scenario when a GPU suddenly becomes unavailable?

**Comment by soulseen** [comment]
yes, the GPU hotplug/unplug lifecycle is independent of the pod/containers. We have some applications that only need the GPU at certain times, and for the times when the GPU is not needed, we want to be able to hotplug/unplug the GPU to other applications.

**Comment by soulseen** [comment]
yes, as you said, in the view of decoupling, GPU resources will be independent of k8s,  and they will be controlled by other components that will hotplug GPU to applications.

**Comment by soulseen** [comment]
@fengwang666 sounds good, But this approach will cause the pod/container to be recreated, this is not expected.

**Comment by zvonkok** [comment]
@soulseen Can you elaborate more on the use-case, as @bergwolf already mentioned, this is not something Kubernetes can achieve and will break the semantics of how devices are handled in a container. What is the concern here? Have a look at our recent blog post on how to improve GPU utilization in Kubernetes: https://developer.nvidia.com/blog/improving-gpu-utilization-in-kubernetes/
BTW. we also added a new resource management API to Kube please take a  look: https://github.com/kubernetes/enhancements/pull/3064

**Comment by bergwolf** [comment]
@zvonkok Thanks for the links to KEP-3063! It would be great to have something flexible like KEP-3063. One quick question, so with KEP-3063, is the CDI the only interface that kata-containers need to work with? I'm thinking we should start to integrate with CDI to prepare for it.

**Comment by zvonkok** [comment]
@bergwolf Yes CDI would be the low-level building block for KEP-3063. BTW https://github.com/kata-containers/kata-containers/pull/4699 adds support for CDI :)
